src/lib/Algebra.hs

```diff
@@ -20,81 +18,101 @@ import Data.Map.Strict hiding (foldl, map, empty, (!))
 import Data.Text.Prettyprint.Doc
 import Data.List (intersperse)
 import Data.Tuple (swap)
-import Data.Coerce
 
-import Builder hiding (sub, add)
+import Builder hiding (sub, add, mul)
 import Simplify
 import Syntax
-import QueryType
 import Name
 import Err
 import MTL1
+import QueryType
 
 type PolyName = Name AtomNameC
 type PolyBinder = NameBinder AtomNameC
 
--- Set of monomials, each multiplied by a constant.
-type Constant         = Rational
-newtype PolynomialP (mono::E) (n::S) =
-  Polynomial { fromPolynomial :: Map (mono n) Constant }
-  deriving (Show, Eq, Ord)
+type Constant = Rational
 
 -- Set of variables, each with its power.
 data Monomial n = Monomial { fromMonomial :: Map (PolyName n) Int }
                   deriving (Show, Eq, Ord)
-type Polynomial = PolynomialP Monomial
-
--- Clamp represents the expression max(poly, 0).
-newtype Clamp (n::S) = Clamp (Polynomial n)
-        deriving (Show, Eq, Ord, HoistableE, SinkableE, AlphaEqE)
 
--- TODO: Make clamps a map of powers!
--- A product of a number of clamps and a monomial.
-data ClampMonomial   (n::S) = ClampMonomial [Clamp n] (Monomial n) deriving (Show, Eq, Ord)
-type ClampPolynomial = PolynomialP ClampMonomial :: E
-
--- === Polynomial math ===
-
-mulC :: ClampPolynomial n-> ClampPolynomial n -> ClampPolynomial n
-mulC (Polynomial x) (Polynomial y) =
-  cpoly [ (cx * cy, mulMono mx my)
-        | (mx, cx) <- toList x, (my, cy) <- toList y]
-  where mulMono (ClampMonomial cx (Monomial mx))
-                (ClampMonomial cy (Monomial my)) =
-          ClampMonomial (cx ++ cy) (Monomial $ unionWith (+) mx my)
+-- Set of monomials, each multiplied by a constant.
+newtype Polynomial (n::S) =
+  Polynomial { fromPolynomial :: Map (Monomial n) Constant }
+  deriving (Show, Eq, Ord)
 
-add :: OrdE mono => PolynomialP mono n -> PolynomialP mono n -> PolynomialP mono n
+-- This is the main entrypoint. Doing polynomial math sometimes lets
+-- us compute sums in closed form. This tries to compute
+-- `\sum_{i=0}^(lim-1) body`. `i`, `lim`, and `body` should all have type `Nat`.
+sumUsingPolys :: (Builder m, Fallible1 m, Emits n)
+              => Atom n -> Abs Binder Block n -> m n (Atom n)
+sumUsingPolys lim (Abs i body) = do
+  Abs i' body' <- refreshAbs (Abs i body) \i' body' ->
+    Abs i' <$> simplifyBlockToBlock body'
+  Abs i'' body'' <- hoistDecls i' body'
```

Comment:
I feel like both simplification and this hoisting should not be the responsibility of algebra